src/Algebra/Graph.hs

```diff
@@ -1066,3 +1105,73 @@ sparsify graph = res
         m <- get
         put (m + 1)
         overlay <$> s `x` m <*> m `y` t
+
+{- Note [The rules of foldg]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The rules for foldg work like this.
+
+* Up to (but not including) phase 1, we use the "buildG/f" rule to
+  rewrite all saturated applications of f with its buildG/foldg
+  form, hoping for fusion to happen (through the "foldg/buildG" rule).
+
+  In phase 1 and 0, we switch off these rules, inline buildG, and
+  switch on the "graph/f" rule, which rewrites the "foldg/f"
+  thing back into plain map.
+
+  It's important that these two rules aren't both active at once
+  (along with build's unfolding) else we'd get an infinite loop
+  in the rules. Hence the activation control below.
+
+* R functions (R for Rewrite) are here to remember f after applying a
+  "builG/f" rule. These functions are higher-order functions and therefore
+  benefits from inlining.
+
+* The "mapR/mapR" rule optimises compositions of fmapR
+-}
+
+buildG :: forall a. F.Fold a -> Graph a
+buildG = F.foldg Empty Vertex Overlay Connect
+{-# INLINE [1] buildG #-}
+
+-- R functions
+composeR :: (b -> c) -> (a -> b) -> a -> c
+composeR = (.)
+{-# INLINE [0] composeR #-}
+
+matchR :: b -> (a -> b) -> (a -> Bool) -> a -> b
+matchR e v p = \x -> if p x then v x else e
+{-# INLINE [0] matchR #-}
+
+flipR :: (b -> b -> b) -> (b -> b -> b)
+flipR = flip
+{-# INLINE [0] flipR #-}
+
+-- Rules to transform a buildG-equivalent function into its equivalent
+{-# RULES
+-- Transform a fmapR into its build equivalent
+"buildG/map"       [~1] forall f g.
+  fmapR   f g  = buildG (F.Fold $ \e v o c -> foldg e (composeR v f) o c g)
+
+-- Transform a induce into its build equivalent
+"buildG/induce"    [~1] forall p g.
+  induce p g  = buildG (F.Fold $ \e v o c -> foldg e (matchR e v p) o c g)
+
+"buildG/transpose" [~1] forall g.
+  foldg Empty Vertex Overlay (flipR Connect) g = buildG (F.Fold $ \e v o c -> foldg e v o (flip c) g)
+ #-}
+
+-- Rules to merge rewrited functions
+{-# RULES
+-- Merge a foldg followed by a buildG
+"foldg/buildG" forall e v o c (g::F.Fold a).
+                              foldg e v o c (buildG g) = F.foldg e v o c g
+
+-- Merge two mapR
```

Comment:
I don't understand this comment and/or rule. It doesn't merge two maps, it reduces a 3-way function composition to a 2-way function composition.

Reply:
(I think the rule is fine, is just a misleading comment and name of the rule.)

Reply:
I will update this